File: Python/subsetsum.py
import pandas as pd
import time

def subset_sum(numbers, target):
    solutions = []

    def search(index, current_sum, current_list):

        if current_sum == target:
            if current_list not in solutions:
                solutions.append(current_list[:])
            return

        if index >= len(numbers) or current_sum > target:
            return

        current_num = numbers[index]
        current_list.append(current_num)

        search(index + 1, current_sum + current_num, current_list)

        current_list.pop()

        search(index + 1, current_sum, current_list)

        return

    search(0, 0, [])
    return solutions

def execute_file(sample_path, result_path):

    df = pd.read_excel(result_path, engine='openpyxl')

    language_list = []
    input_size_list = []
    target_list = []
    execution_time_list = []


    with open(sample_path) as current_file:
        data = current_file.readlines()
        
        for i in range(int(len(data) / 3)):
            target = int(data[i*3][:-1])
            numbers = []
            for j in data[i*3 + 1][:-1].split(" "):
                numbers.append(int(j))

            language_list.append("Python")
            input_size_list.append(len(numbers))
            target_list.append(target)

            start = time.perf_counter()
            solutions = subset_sum(numbers, target)
            execution_time_list.append(f"{((time.perf_counter() - start)*1000):.4f}")
    
    df_new = pd.DataFrame({
        'language': language_list,
        'input_size': input_size_list,
        'target': target_list,
        'execution_time': execution_time_list
    })

    if not df.empty:
        df_final = pd.concat([df, df_new], ignore_index=True)
    else:
        df_final = df_new


    df_final['execution_time'] = df_final['execution_time'].astype(float).map('{:.4f}'.format)

    df_final.to_excel(result_path, index=False, engine='openpyxl')
    print(f"Dados de '{sample_path}' processados.")
        
execute_file(".\\Input\\small_input.txt", ".\\Results\\results.xlsx")
execute_file(".\\Input\\med_input.txt", ".\\Results\\results.xlsx")
execute_file(".\\Input\\big_input.txt", ".\\Results\\results.xlsx")

File: Java/SubsetSum.java
package Java;

import org.apache.poi.ss.usermodel.*;
import org.apache.poi.xssf.usermodel.XSSFWorkbook;

import java.io.*;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

public class SubsetSum {

    static List<List<Integer>> solutions = new ArrayList<>();

public static void main(String[] args) {
        String inputFolder = "Input/";   
        String resultsFile = "Results/results.xlsx"; 

        executeFile(inputFolder + "small_input.txt", resultsFile);
        executeFile(inputFolder + "med_input.txt", resultsFile);
        executeFile(inputFolder + "big_input.txt", resultsFile);
    }

    public static List<List<Integer>> subsetSum(List<Integer> numbers, int target) {
        solutions = new ArrayList<>();
        search(0, 0, new ArrayList<>(), numbers, target);
        return solutions;
    }

    private static void search(int index, int currentSum, List<Integer> currentList, List<Integer> numbers, int target) {
        if (currentSum == target) {
            if (!solutions.contains(currentList)) {
                solutions.add(new ArrayList<>(currentList));
                return;
            }
        }

        if (index >= numbers.size() || currentSum > target) {
            return;
        }

        int currentNum = numbers.get(index);
        currentList.add(currentNum);

        search(index + 1, currentSum + currentNum, currentList, numbers, target);

        currentList.remove(currentList.size() - 1);

        search(index + 1, currentSum, currentList, numbers, target);
    }

    public static void executeFile(String samplePath, String resultPath) {
        List<String> languageList = new ArrayList<>();
        List<Integer> inputSizeList = new ArrayList<>();
        List<Integer> targetList = new ArrayList<>();
        List<String> executionTimeList = new ArrayList<>();

        try {
            List<String> data = Files.readAllLines(Paths.get(samplePath));

            for (int i = 0; i < data.size() / 3; i++) {
                try {
                    String targetLine = data.get(i * 3).trim();
                    if (targetLine.isEmpty() || targetLine.equals("---")) continue;

                    int target = Integer.parseInt(targetLine);
                    
                    String numbersLine = data.get(i * 3 + 1).trim();
                    List<Integer> numbers = new ArrayList<>();
                    
                    if (!numbersLine.isEmpty()) {
                        for (String s : numbersLine.split(" ")) {
                            numbers.add(Integer.parseInt(s));
                        }
                    }

                    languageList.add("Java");
                    inputSizeList.add(numbers.size());
                    targetList.add(target);

                    long start = System.nanoTime();
                    subsetSum(numbers, target);
                    long end = System.nanoTime();

                    double durationMs = (end - start) / 1_000_000.0;
                    executionTimeList.add(String.format(Locale.US, "%.4f", durationMs));

                } catch (NumberFormatException | IndexOutOfBoundsException e) {
                }
            }

            File file = new File(resultPath);
            Workbook workbook;
            Sheet sheet;

            if (file.exists()) {
                FileInputStream fis = new FileInputStream(file);
                workbook = new XSSFWorkbook(fis);
                sheet = workbook.getSheetAt(0);
                fis.close();
            } else {
                workbook = new XSSFWorkbook();
                sheet = workbook.createSheet("Results");
                Row header = sheet.createRow(0);
                header.createCell(0).setCellValue("language");
                header.createCell(1).setCellValue("input_size");
                header.createCell(2).setCellValue("target");
                header.createCell(3).setCellValue("execution_time");
            }

            int rowNum = sheet.getLastRowNum() + 1;

            for (int j = 0; j < languageList.size(); j++) {
                Row row = sheet.createRow(rowNum++);
                row.createCell(0).setCellValue(languageList.get(j));
                row.createCell(1).setCellValue(inputSizeList.get(j));
                row.createCell(2).setCellValue(targetList.get(j));

                row.createCell(3).setCellValue(executionTimeList.get(j));
            }

            FileOutputStream fos = new FileOutputStream(file);
            workbook.write(fos);
            workbook.close();
            fos.close();

        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}